src/hotspot/share/gc/shared/stringdedup/stringDedupTable.cpp

```diff
@@ -395,32 +395,33 @@ bool StringDedupTable::is_rehashing() {
 StringDedupTable* StringDedupTable::prepare_resize() {
   size_t size = _table->_size;
 
-  // Check if the hashtable needs to be resized
+  // Decide whether to resize, and compute desired new size if so.
   if (_table->_entries > _table->_grow_threshold) {
-    // Grow table, double the size
-    size *= 2;
-    if (size > _max_size) {
-      // Too big, don't resize
-      return NULL;
+    // Compute new size.
+    size_t needed = _table->_entries / _grow_load_factor;
+    if (needed < _max_size) {
+      size = round_up_power_of_2(needed);
+    } else {
+      size = _max_size;
     }
   } else if (_table->_entries < _table->_shrink_threshold) {
-    // Shrink table, half the size
-    size /= 2;
-    if (size < _min_size) {
-      // Too small, don't resize
-      return NULL;
-    }
-  } else if (StringDeduplicationResizeALot) {
-    // Force grow
-    size *= 2;
-    if (size > _max_size) {
-      // Too big, force shrink instead
-      size /= 4;
+    // Compute new size.  We can't shrink by more than a factor of 2,
+    // because the partitioning for parallelization doesn't support more.
+    if (size > _min_size) size /= 2;
+  }
+  // If no change in size needed (and not forcing resize) then done.
+  if (size == _table->_size) {
+    if (!StringDeduplicationResizeALot) {
+      return NULL;              // Don't resize.
+    } else if (size < _max_size) {
+      size *= 2;                // Force grow, but not past _max_size.
+    } else {
+      size /= 2;                // Can't force grow, so force shrink instead.
     }
-  } else {
-    // Resize not needed
-    return NULL;
   }
+  assert(size <= _max_size, "invariant: %zu", size);
+  assert(size >= _min_size, "invariant: %zu", size);
+  assert(is_power_of_2(size), "invariant: %zu", size);
```

Comment:
Please also use `SIZE_FORMAT` instead of `%zu` like in the overwhelming majority for formatting `size_t`.

Reply:
I actually prefer `%zu`. I can't seem to find whether it's permitted or not in HotSpot Coding Style though.

Reply:
This was discussed some time ago (maybe only Oracle internal?).  So far as I can recall, everyone preferred using the 'z' qualifier and nobody liked the macro.  The only concern was backports, but that didn't seem to be a problem.  The consensus was to use it in new code and file an RFE for updating existing uses:
https://bugs.openjdk.java.net/browse/JDK-8256379